### PR TITLE
CI: Run logformatter for podman_machine_windows_task

### PR DESCRIPTION
**Pull request by l0rd**
That's to address the windows part of https://github.com/containers/podman/issues/21760

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

File: contrib/cirrus/win-podman-machine-test.ps1 [hunk 1]
```diff
@@ -33,5 +33,12 @@ Run-Command ".\bin\windows\podman.exe --version"
 New-Item -ItemType "directory" -Path "$env:AppData\containers"
 Copy-Item -Path pkg\machine\ocipull\policy.json -Destination "$env:AppData\containers"
 
+# Install Perl
+Write-Host "`nInstalling Perl as it is required to use logformatter"
+Run-Command 'choco install --no-progress --confirm --acceptlicense --nocolor StrawberryPerl'
```
Comment (Luap99) [inline contrib/cirrus/win-podman-machine-test.ps1:38]: I guess that should be moved to our image builds: https://github.com/containers/automation_images/blob/a3e4099c72052ff21a6e538043cf05b5d873a110/win_images/win_packaging.ps1#L12

File: contrib/cirrus/win-podman-machine-test.ps1 [hunk 2]
```diff
@@ -33,5 +33,12 @@ Run-Command ".\bin\windows\podman.exe --version"
...
+Write-Host "`nAdd perl to the PATH"
+Set-Item "Env:PATH" "$Env:PATH;C:\Strawberry\perl\bin"
+
+# Run the tests
 Write-Host "`nRunning podman-machine e2e tests"
-Run-Command ".\winmake localmachine"
+Run-Command '.\winmake localmachine | perl .\contrib\cirrus\logformatter $Env:TEST_FLAVOR-podman-windows-rootless-host-sqlite'
```
Comment (Luap99) [inline contrib/cirrus/win-podman-machine-test.ps1:44]: I have no idea how windows deals with this but will the overall command still fail when the test suite fails?
I am a bit paranoid here, can you make a negative test here, i.e. add commit with triggers an intentional failure in the tests to make sure the CI job is still marked as failed in this case.

**Review by Luap99 — COMMENTED** [review]
Thanks, logs look much better with that

File: contrib/cirrus/win-podman-machine-test.ps1 (hunk 2 above)
Comment (l0rd) [reply contrib/cirrus/win-podman-machine-test.ps1:44]: Good idea. Let me try that.

Comment (l0rd) [reply contrib/cirrus/win-podman-machine-test.ps1:44]: If one test fails, the CI job is correctly marked as failed. But the html log file is not generated. I will figure out why.

**Comment by edsantiago** [comment]
Thank you! I'm working on logformatter, adding code to recognize `C:` and backslashes and that sort of thing.

One request: could you find some way to inject `/define.gitCommit=XXXXX` into the log output? Without that there's no way to linkify back to source code. Here's how I did it for Macintosh: https://github.com/containers/podman/blob/8a643c243ea8a9f9ac95ed137d014b90797470ca/Makefile#L632-L636

...but of course that relies on an invocation of the form `make localmachine |& logformatter`

**Comment by edsantiago** [comment]
@l0rd feel free to grab #22081

File: contrib/cirrus/win-podman-machine-test.ps1 (hunk 1 above)
Comment (edsantiago) [reply contrib/cirrus/win-podman-machine-test.ps1:38]: Done but untested: https://github.com/containers/automation_images/pull/337#issuecomment-2005057595

**Comment by l0rd** [comment]
> One request: could you find some way to inject /define.gitCommit=XXXXX into the log output? Without that there's no way to linkify back to source code.

That should not be a problem assumed that the env variable `$GIT_COMMIT` is set on windows too. I will look at it.

> @l0rd feel free to grab https://github.com/containers/podman/pull/22081

👍 I am rebasing my PR to grab it

**Comment by l0rd** [comment]
I have update the PR so that the html report is generated when tests fail too (this was tricky because there is no way to set `pipefail` with powershell as we do on bash scripts).

Also now the reports first line has the git commit: `/define.gitCommit=<SHA>`

Now the windows cirrus tasks install Perl if it's not installed yet. We can merge the PR as is or wait for the new VMs with Perl and remove the code that installs Perl from this PR. 

Tests failing examples:
- [machine-wsl](https://api.cirrus-ci.com/v1/artifact/task/6383773640556544/html/repo/machine-wsl-podman-windows-rootless-host-sqlite.log.html) ([cirrus dashboard](https://cirrus-ci.com/task/6383773640556544))
- [machine-hyperv](https://api.cirrus-ci.com/v1/artifact/task/4976398757003264/html/repo/machine-hyperv-podman-windows-rootless-host-sqlite.log.html) ([cirrus dashboard](https://cirrus-ci.com/task/4976398757003264))

Tests passing examples:
- [machine-wsl](https://api.cirrus-ci.com/v1/artifact/task/4568560064790528/html/repo/machine-wsl-podman-windows-rootless-host-sqlite.log.html) ([cirrus dashboard](https://cirrus-ci.com/task/4568560064790528))
- [machine-hyperv](https://api.cirrus-ci.com/v1/artifact/task/5694459971633152/html/repo/machine-hyperv-podman-windows-rootless-host-sqlite.log.html) ([cirrus dashboard](https://cirrus-ci.com/task/5694459971633152))

File: .cirrus.yml [hunk 3]
```diff
@@ -791,6 +791,11 @@ podman_machine_windows_task:
           TEST_FLAVOR: "machine-hyperv"
     clone_script: *winclone
     main_script: ".\\repo\\contrib\\cirrus\\win-podman-machine-main.ps1"
+    always:
+        # Required for `contrib/cirrus/logformatter` to work properly
+        html_artifacts:
+            path: ./repo/*.html
```
Comment (edsantiago) [inline .cirrus.yml:797]: The logs look great! Thank you! But this spurious `/repo` breaks all existing conventions, and makes the logs impossible to find. Is there any way to remove the `/repo`?

**Comment by edsantiago** [comment]
Also, you might not know about this: https://github.com/edsantiago/greasemonkey/tree/master/github-ci-highlight

In addition to colorizing github jobs in a way that makes patterns stand out, it also auto-links to colorized logs, adding a single-click trivial way to jump straight to logs. With the `/repo` thing, this linkification no longer works. Nor does my flake logger.

File: .cirrus.yml (hunk 3 above)
Comment (edsantiago) [reply .cirrus.yml:797]: Sigh. Doesn't look like there's a way to do that [with Cirrus](https://cirrus-ci.org/guide/writing-tasks/#artifacts-instruction). Can the file be written (or moved) to `..`?

Comment (edsantiago) [reply .cirrus.yml:797]: Without a 1-click link the formatted logs are not really very useful. However, it's still incremental progress and maybe other Windows peeps will want it.

**Review by edsantiago — APPROVED** [review]

**Comment by l0rd** [comment]
> Also, you might not know about this: [edsantiago/greasemonkey@`master`/github-ci-highlight](https://github.com/edsantiago/greasemonkey/tree/master/github-ci-highlight?rgh-link-date=2024-03-26T16%3A24%3A48Z)
> 
> In addition to colorizing github jobs in a way that makes patterns stand out, it also auto-links to colorized logs, adding a single-click trivial way to jump straight to logs. With the `/repo` thing, this linkification no longer works. Nor does my flake logger.

Thanks for sharing. I was not aware of it. 

I made an update to the PR to move the report one folder up and now the link works.

**Comment by edsantiago** [comment]
Confirmed, the link now works. Thank you!

File: contrib/cirrus/win-podman-machine-test.ps1 (hunk 2 above)
Comment (l0rd) [reply contrib/cirrus/win-podman-machine-test.ps1]: Resolving as the problem has been addressed. The output of the command is always captured in an env variable and processed by `logformatter`. After that an exception is thrown if the (native) command exit code is not zero (same behavior as before).

File: contrib/cirrus/win-podman-machine-test.ps1 (hunk 1 above)
Comment (l0rd) [reply contrib/cirrus/win-podman-machine-test.ps1]: Currently we are checking if Perl exists before invoking logformatter and install it if necessary:

https://github.com/containers/podman/blob/fb046fc24a16af87357a7f1335661c88691e2f40/contrib/cirrus/win-lib.ps1#L43-L50

I will open another PR to remove this code as soon as Perl is available in the new windows image.

Comment (edsantiago) [reply contrib/cirrus/win-podman-machine-test.ps1]: #22248

**Pull request merged**